Support `BuildSpec` in `BuildRun` resources

HeavyWombat opened the pull request:
# Changes

Fixes #677

Based upon [SHIP #29](https://github.com/shipwright-io/community/blob/main/ships/0029-one-off-builds.md)

Core SHIP proposals:

- [X] Have `BuildSpec` in `BuildRun`
- [X] Embedded build spec does not lead to `Build` registered in the cluster.
- [X] Validation steps are done against embedded Build(Spec).
- [X] The embedded build spec, does not count as a Build in the context of the metrics package.
- [X] No TaskRun is created when Build(Spec) is invalid
- [X] BuildSpec is not used in combination with `BuildRef`, `output`, `paramValues`, `env`, or `timeout`.
- [X] README is updated with embedded example instead of Build and BuildRun example

# Submitter Checklist

- [X] Includes tests if functionality changed/was added
- [X] Includes docs if changes are user-facing
- [X] [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)  
- [X] Release notes block has been filled in, or marked NONE

See [the contributor guide](https://github.com/shipwright-io/build/blob/main/CONTRIBUTING.md)
for details on coding conventions, github and prow interactions, and the code review process.

# Release Notes

```release-note
Introducing support to embed a BuildSpec inside a BuildRun to have one-off builds, where only a BuildRun is required without the need of a Build resource. This includes an API change as the BuildRef in BuildRuns is no longer mandatory. Either BuildRef or BuildSpec can be used.
```

HeavyWombat commented:
I only tested on Kube `1.21` so far, the E2E tests are failing in a `1.20` setup. Have to have a look at this.

SaschaSchwarze0 commented:
> I only tested on Kube `1.21` so far, the E2E tests are failing in a `1.20` setup. Have to have a look at this.

No worries @HeavyWombat, that looked like the typical hanging BuildAh build.

SaschaSchwarze0 commented on pkg/reconciler/buildrun/resources/sources/http.go:
```diff
@@ -25,7 +25,7 @@ func AppendHTTPStep(
 	httpStep := findExistingHTTPSourcesStep(taskSpec)
 	if httpStep != nil {
-		httpStep.Container.Args[3] = fmt.Sprintf("%s ; wget \"%s\"", httpStep.Container.Args[3], source.URL)
+		httpStep.Container.Args[3] = fmt.Sprintf("%s ; wget %q", httpStep.Container.Args[3], source.URL)
```
Fix is nice. Potentially even fixing a code injection. But why not in a separate PR? Simlar concern applies to your restructuring in the build reconciler. This is not nice for reviewers.

HeavyWombat replied:
No problem creating a spin-off based on this commit.

SaschaSchwarze0 commented on pkg/validate/validate.go:
```diff
@@ -67,3 +67,42 @@ func NewValidation(
+		}
+
+		if buildRun.Spec.Output != nil {
+			return fmt.Errorf("cannot use 'output' override and 'buildSpec' simultaneously")
```
Looking for feedback from a native speaker. I think "simultaneously" is more meant for two different activities that happen at the some time. Here, "together" might be a better fit.

SaschaSchwarze0 reviewed (changes requested):
Nice work. The pragmatic way of using an inline build has advantages and disadvantages. Without encouraging you to go away from it, there is one thing which imo should not happen: at certain places, we store a Build name: as a label on the BuildRun or TaskRun. And as label in Prometheus metrics. I don't think we should store a fake value there.

Imo BuildRun and TaskRun labels should not be set at all. In Prometheus, we must provide a value for the build label, but maybe `""` is a better choice.

dalbar commented on pkg/reconciler/buildrun/resources/build.go:
```diff
@@ -8,27 +8,43 @@ import (
 
-	return err
+	// Bail out hard in case of an invalid state (unreachable code)
+	panic("invalid BuildRun resource that neither has a build reference nor an embedded build specification")
```
I don't think that a "getter" function should be unsafe and possibly crash the entire controller. Since the api of the function already offers to return an error, I would like to see an error for the panic case that is handled higher up.

dalbar commented on pkg/apis/build/v1alpha1/buildrun_types.go:
```diff
@@ -351,3 +364,18 @@ func (brs *BuildRunStatus) SetCondition(condition *Condition) {
+	}
+
+	// unreachable code for properly validated BuildRun resources
+	panic("invalid BuildRunSpec with neither BuildRef nor BuildSpec")
```
Why can't we rather return an empty string over here? We know that a name can never be empty so the semantics should be clear. The function would also be "safe" to run in a vacuum. The panic seems out-of-scope for this function.

HeavyWombat replied:
I am also fine with an empty string.

dalbar reviewed (changes requested):
Great work @HeavyWombat especially on the validation and tests. I have some minor change requests.

HeavyWombat commented on README.md:
````diff
@@ -63,54 +63,32 @@ Shipwright supports any tool that can build container images in Kubernetes clust
   ```
 
-* Create a *Build* object, replacing `<REGISTRY_ORG>` with the registry username your `push-secret` secret have access to:
+- Create and submit your *BuildRun* object, replacing `<REGISTRY_ORG>` with the registry username your `push-secret` secret have access to:
````
Yeah, I was thinking about this one, because there were some linter issues I wanted to resolve anyway. So my rational was: The example in the README is a quick try it out kind of example. So it would be nice if it would be as simple as possible. What do you think?

SaschaSchwarze0 replied:
It should be simple, I agree. But, it also is the first sample that our users might see. If we show the BuildRun-only use case, then I think we must at least make clear that this is not the only possibility but maybe used as the simplest option to verify the installation.

qu1queee replied:
Plus one on not changing this. I prefer to introduce a sentence to highlight on this other feature or to extend into a new tutorial using the embedded `BuildSpec`. The design of SHP is based on two core resources, this is what we should pitch on our landing document, as using a Build + BuildRun have a larger scope in terms of use cases than the one-off Build.

HeavyWombat replied:
Removed.

HeavyWombat commented on pkg/reconciler/buildrun/buildrun.go:
```diff
@@ -135,9 +139,49 @@ func (r *ReconcileBuildRun) Reconcile(ctx context.Context, request reconcile.Req
+
+				default:
+					// TODO error or panic
+					panic("invalid BuildRun, neither BuildRef nor BuildSpec are defined")
```
Actually, this part of the code should not be reached, since the previous validation already caught this.

HeavyWombat commented on docs/buildrun.md:
````diff
@@ -80,6 +81,31 @@ spec:
     name: buildpack-nodejs-build-namespaced
 ```
 
+_Please note:_ The `BuildRef` and `BuildSpec` is mutually exclusive.
````
Will be fixed.

HeavyWombat commented on docs/buildrun.md:
````diff
@@ -80,6 +81,31 @@ spec:
+      image: foo/bar:latest
+```
+
+_Please note:_ The `BuildRef` and `BuildSpec` is mutually exclusive.
````
Will be fixed.

qu1queee reviewed (changes requested):
Thanks this is nice. From my review I see the following:

[1] Introducing too many changes not related to this feature. Some of them are nice fixes, the rest I do not understand the rational, but they can go in. I'm just concerned that your code style is not gonna get widely adopted, so we will see what you change today again the future(take the imports as an example), not sure how much of value it have then.  With this in mind, pls  squash the following commits into one:
- Fix potentially unused variable warning
- Sort imports in build_test.go
- Use specific description for bundle test
- Add convenience checks for test env vars
- Fix duplicate import
- Refactor to fix potential variable shadow issue
- Fix build not found test case
- Clean-up Makefile
- Fix declaration of "err" shadows declaration issue
- Simplify if error case
- Add/remove empty lines in buildrun_test.go

[2] I wrote a comment on this pr reintroducing a pointer. I was not involved on the discussion but we have that pointer before and then we went through a long discussion and that was removed. Wondering if you are aware of this.

qu1queee commented on pkg/reconciler/buildrun/buildrun.go:
```diff
@@ -101,17 +100,24 @@ func (r *ReconcileBuildRun) Reconcile(ctx context.Context, request reconcile.Req
+			r.client,
+			buildRun,
+			err.Error(),
+			resources.BuildRunInvalid,
```
+1

qu1queee commented on pkg/reconciler/buildrun/buildrun.go:
```diff
@@ -128,9 +139,48 @@ func (r *ReconcileBuildRun) Reconcile(ctx context.Context, request reconcile.Req
+
+				// When the build(spec) is embedded in the buildrun, the now
+				// transient/volatile build resource needs to be validated first
+				case buildRun.Spec.BuildSpec != nil:
```
this is neat!

qu1queee commented on docs/buildrun.md:
```diff
@@ -80,6 +81,31 @@ spec:
 
+_Please note:_ The `BuildRef` and `BuildSpec` are mutually exclusive.
+
+### Defining the BuildSpec
```
pls add this to the top index

qu1queee replied:
@HeavyWombat this still missing.

HeavyWombat replied:
@SaschaSchwarze0 @qu1queee Not sure why this is not marked as outdated here, but I moved the note to the common section: https://github.com/shipwright-io/build/pull/1016/commits/e735e50e5447bbe99243dbbbbd8540c795371076.

qu1queee replied:
@HeavyWombat I think I was not clear enough. You need to add the new heading level into the [index](https://github.com/shipwright-io/build/pull/1016/files#diff-a4578c2cbd97c42d96f1dc937794172a093bcd676f6d6f190b7452f54539996eR11-R14)

HeavyWombat replied:
Yeah, I misunderstood that. I am sorry. It should be fixed now.

qu1queee commented on pkg/reconciler/build/build.go:
```diff
@@ -22,6 +22,17 @@ import (
 )
 
+// build a list of current validation types
+var validationTypes = [...]string{
```
I think we need to align the order of validations here with the one on the Build controller. For example, if we have the wrong strategy and the wrong secret on a Build and on a one-off BuildRun, the error reported on the status sub resource will be different. One will point out to a wrong secret, the other to a wrong strategy, we want both to say the same.

HeavyWombat replied:
Good catch, I fixed the order to match the sequence.

HeavyWombat commented on pkg/apis/build/v1alpha1/buildrun_types.go:
```diff
@@ -23,7 +23,9 @@ const (
-	BuildRef BuildRef `json:"buildRef"`
+	//
+	// +optional
+	BuildRef *BuildRef `json:"buildRef"`
```
No, this is fine. With the introduction of `BuildSpec`, the `BuildRef` needs to be optional now. Otherwise we could have both a build reference and an embedded build(spec).

HeavyWombat commented:
> Thanks this is nice. From my review I see the following:
> 
> [1] Introducing too many changes not related to this feature. Some of them are nice fixes, the rest I do not understand the rational, but they can go in. I'm just concerned that your code style is not gonna get widely adopted, so we will see what you change today again the future(take the imports as an example), not sure how much of value it have then. With this in mind, pls squash the following commits into one:
> 
>     * Fix potentially unused variable warning
> 
>     * Sort imports in build_test.go
> 
>     * Use specific description for bundle test
> 
>     * Add convenience checks for test env vars
> 
>     * Fix duplicate import
> 
>     * Refactor to fix potential variable shadow issue
> 
>     * Fix build not found test case
> 
>     * Clean-up Makefile
> 
>     * Fix declaration of "err" shadows declaration issue
> 
>     * Simplify if error case
> 
>     * Add/remove empty lines in buildrun_test.go
> 
> 
> [2] I wrote a comment on this pr reintroducing a pointer. I was not involved on the discussion but we have that pointer before and then we went through a long discussion and that was removed. Wondering if you are aware of this.

This work resulted in a number of related side changes. They added up over time to be honest. These were only done in files that I had to touch anyway. For consistency, I updated the same for example linter issue in the same file, too. Where possible, I opted to create a spin-off PR and moved the commit into it, i.e. https://github.com/shipwright-io/build/pull/1018 and https://github.com/shipwright-io/build/pull/1019.

With regards to the sorting order of imports, @gabemontero please keep me honest, I think we spoke about this in an earlier PR. The idea would be to have the Kubeish imports separately. I kind of like the idea, helps with getting a quick overview. However, we have no automated way to "enforce" the order. Some files have it, some do not. In come cases, other PRs updated the imports to have them sorted. If there is some easy and automated way to have that, we maybe should follow-up on that.

qu1queee commented:
@HeavyWombat pls see the comment on squashing

> pls squash the following commits into one

HeavyWombat commented:
> @HeavyWombat pls see the comment on squashing
> 
> > pls squash the following commits into one

@qu1queee I fail to see the rational behind this to be honest. To squash them into one commit, I have to come up with a generic subject line for all of the combined changes. This will be very generic description. If I for example use GitLens, rather than a very specific subject, I only see the generic "fixing things" subject. Care to share your thoughts behind the squashing?

HeavyWombat commented:
> Nice work. The pragmatic way of using an inline build has advantages and disadvantages. Without encouraging you to go away from it, there is one thing which imo should not happen: at certain places, we store a Build name: as a label on the BuildRun or TaskRun. And as label in Prometheus metrics. I don't think we should store a fake value there.
> 
> Imo BuildRun and TaskRun labels should not be set at all. In Prometheus, we must provide a value for the build label, but maybe `""` is a better choice.

I added a logic to not set a build name label in case of an embedded buildrun.

SaschaSchwarze0 commented:
> > Nice work. The pragmatic way of using an inline build has advantages and disadvantages. Without encouraging you to go away from it, there is one thing which imo should not happen: at certain places, we store a Build name: as a label on the BuildRun or TaskRun. And as label in Prometheus metrics. I don't think we should store a fake value there.
> > Imo BuildRun and TaskRun labels should not be set at all. In Prometheus, we must provide a value for the build label, but maybe `""` is a better choice.
> 
> I added logic to not set a build name label in case of an embedded buildrun.

@HeavyWombat may you check here: https://github.com/shipwright-io/build/blob/35d8e003f60fda781df182c0a730e86af6a2f4f8/pkg/reconciler/buildrun/buildrun.go#L234. Is not this `if` loop entered?

HeavyWombat commented:
> > > Nice work. The pragmatic way of using an inline build has advantages and disadvantages. Without encouraging you to go away from it, there is one thing which imo should not happen: at certain places, we store a Build name: as a label on the BuildRun or TaskRun. And as label in Prometheus metrics. I don't think we should store a fake value there.
> > > Imo BuildRun and TaskRun labels should not be set at all. In Prometheus, we must provide a value for the build label, but maybe `""` is a better choice.
> > 
> > I added logic to not set a build name label in case of an embedded buildrun.
> 
> @HeavyWombat may you check here: https://github.com/shipwright-io/build/blob/35d8e003f60fda781df182c0a730e86af6a2f4f8/pkg/reconciler/buildrun/buildrun.go#L234. Is not this `if` loop entered?

Thanks, I will check.

dalbar reviewed (approved):
Thank you for addressing the change requests and removing the panics. Great contribution.

HeavyWombat commented:
> > > > Nice work. The pragmatic way of using an inline build has advantages and disadvantages. Without encouraging you to go away from it, there is one thing which imo should not happen: at certain places, we store a Build name: as a label on the BuildRun or TaskRun. And as label in Prometheus metrics. I don't think we should store a fake value there.
> > > > Imo BuildRun and TaskRun labels should not be set at all. In Prometheus, we must provide a value for the build label, but maybe `""` is a better choice.
> > > 
> > > 
> > > I added logic to not set a build name label in case of an embedded buildrun.
> > 
> > 
> > @HeavyWombat may you check here: https://github.com/shipwright-io/build/blob/35d8e003f60fda781df182c0a730e86af6a2f4f8/pkg/reconciler/buildrun/buildrun.go#L234
> > . Is not this `if` loop entered?
> 
> Thanks, I will check.

@SaschaSchwarze0 Extended the test case to make sure to catch it and modified the section to avoid having the BuildRun update. Thanks for the pointer.

SaschaSchwarze0 commented on docs/buildrun.md:
```diff
@@ -247,13 +271,13 @@ The `status.conditions` hosts different fields, like `status`, `reason` and `mes
+| Status   | Reason                                  | CompletionTime is set | Description |
+| ---      | ---                                     | --- | --- |
+| Unknown  | Pending                                 | No  | The BuildRun is waiting on a Pod in status Pending. |
+| Unknown  | Running                                 | No  | The BuildRun has been validate and started to perform its work. |l
```
```suggestion
| Unknown  | Running                                 | No  | The BuildRun has been validate and started to perform its work. |
```

SaschaSchwarze0 commented on pkg/reconciler/buildrun/buildrun_test.go:
```diff
@@ -1216,6 +1214,271 @@ var _ = Describe("Reconcile BuildRun", func() {
+				})
+
+				_, err := reconciler.Reconcile(context.TODO(), buildRunRequest)
+				Expect(err).To(BeNil())
```
Hm, I just notice that this is missing at many places including existing tests. But don't we need this to verify that the update call happened at all?

```suggestion
				Expect(err).To(BeNil())
				Expect(client.UpdateStatusCallCount()).To(Equal(1))
```

HeavyWombat replied:
I wrote the test cases in such a way that they fail with the `Expect`, so that I know the respective section is hit. But, I agree, it is A) cleaner/more precise to have the check as well and B) helps with potential regressions when common test code is touched. Will add the check

SaschaSchwarze0 reviewed (changes requested):
Code is all fine, unsure about a detail in the tests. A few comments in the documentation are still outstanding.

SaschaSchwarze0 reviewed (approved):
/approve

qu1queee reviewed (approved):
All good from my side, this is good stuff!!

gabemontero commented:
> Similar as in the [other one](https://github.com/shipwright-io/build/pull/1020), given this includes an API change, we'll need a second review. @adambkaplan @gabemontero

I'm good with API ... very similar pattern as to seen in k8s and tekton

I'll 

/lgtm

but will 

/hold 

in case @adambkaplan gets time to take a look and chime in.  If he doesn't by EOB Friday Apr 1, go ahead and unhold.

adambkaplan reviewed (commented):
Generally looks good, I request that you squash commits before merging (all code changes can be bundled together).

gabemontero commented:
/unhold

HeavyWombat commented on pkg/reconciler/buildrun/resources/build.go:
```diff
@@ -8,27 +8,43 @@ import (
+		build.Name = ""
+		build.Namespace = buildRun.Namespace
+		build.Spec = *buildRun.Spec.BuildSpec
+		build.Status = buildv1alpha1.BuildStatus{}
```
Yes, this is [done in the reconciler](https://github.com/shipwright-io/build/pull/1016/files#diff-92a124fc33b3ec7d38392c0b313144f552c7987f35bb1aad07ea355020ad32e3R153-R186) code.

HeavyWombat commented on pkg/reconciler/buildrun/resources/build.go:
```diff
@@ -8,27 +8,43 @@ import (
+	if buildRun.Spec.BuildSpec != nil {
+		build.Name = ""
+		build.Namespace = buildRun.Namespace
+		build.Spec = *buildRun.Spec.BuildSpec
```
Added a deepcopy instead. Thanks for the hint.

HeavyWombat commented on docs/buildrun.md:
```diff
@@ -275,6 +300,9 @@ The following table illustrates the different states a BuildRun can have under i
 | False    | BuildNotFound                           | Yes | The related Build in the BuildRun was not found. |
 | False    | BuildRunCanceled                        | Yes | The BuildRun and underlying TaskRun were canceled successfully. |
 | False    | BuildRunNameInvalid                     | Yes | The defined `BuildRun` name (`metadata.name`) is invalid. The `BuildRun` name should be a [valid label value](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set). |
+| False    | BuildRunNoRefOrSpec                     | Yes | BuildRun does not have either `BuildRef` or `BuildSpec` defined. There is no connection to a Build specification. |
+| False    | BuildRunAmbiguousBuild                  | Yes | The defined `BuildRun` uses both `BuildRef` and `BuildSpec`. Only one of them is allowed at the same time.|
+| False    | BuildRunBuildFieldOverrideForbidden     | Yes | The defined `BuildRun` uses an override (e.g. `timeout`, `paramValues`, `output`, or `env`) in combination with `BuildSpec`, which is not allowed. Use the `BuildSpec` to directly specify the respective value. |
```
Yeah, exactly, I dropped a comment with that in the code at the respective place.

HeavyWombat commented:
Had a bit of a rebase hick-up of sorts. Should be fixed now.

SaschaSchwarze0 commented:
Seeing all comments from Adam addressed, re-adding `lgtm`

/lgtm

qu1queee commented:
/lgtm
